apoc.load.html procedure fail silently

Currently, the `apoc.load.html` procedure returns an error and rolls back the transaction if any error occurs. The error messaging is not really helpful, you can look at https://github.com/neo4j-contrib/neo4j-apoc-procedures/blob/4.1/full/src/main/java/apoc/load/LoadHtml.java#L54.
In my opinion, it would be better to fail silently and just return an empty map with defined keys. Something along the line of:

```
  query.keySet().stream().forEach(key -> {
      output.put(key, empty map);
  });
```

@tomasonjo

[apoc-4.1.0.6.zip](https://github.com/neo4j-contrib/neo4j-apoc-procedures/files/6034758/apoc-4.1.0.6.zip)

I attached a jar with the "fail silently" feature. Now, with config `{failsSilently: true}`, the procedure doesn't fail with exception, but continue and create a Neo4j `log.warn` with the failed tag, for example 
```
WARN  Error during parsing element: <a hre f="#cite_ref-Simpson2018_1-0"><sup><i><b>a</b></i></sup></a>
```
and returns the list of the values, except for values not correctly parsed.
Otherwise, with `{failsSilently: false}` (default value), the procedure fails with an error like this:
```
Failed to invoke procedure `apoc.load.html`: Caused by: java.lang.RuntimeException: Error during parsing element: <a hre f="#cite_ref-Simpson2018_1-0"><sup><i><b>a</b></i></sup></a>
```



I created a pr with this change #1817

--- a/docs/asciidoc/modules/ROOT/partials/usage/apoc.load.html.adoc
+++ b/docs/asciidoc/modules/ROOT/partials/usage/apoc.load.html.adoc
@@ -143,3 +143,85 @@ a|
 }
 ----
 |===
+
+Let's suppose we have a `test.html` file like this:
+[source,html]
+----
+<!DOCTYPE html>
+<html class="client-nojs" lang="en" dir="ltr">
+  <h6 i d="error">test</h6>
+  <h6 id="correct">test</h6>
+</html>
+----
+
+We can handle the parse error caused by `i d` through `failSilently` configuration.
+So, we can execute:
+
+[source,cypher]
+----
+CALL apoc.load.html("test.html",{h6:"h6"});
+----
+.Results
+[opts="header"]
+|===
+| Failed to invoke procedure `apoc.load.html`: Caused by: java.lang.RuntimeException: Error during parsing element: <h6 i d="error">test</h6>
+|===
+
+or with failSilently `WITH_LIST`:
+
+[source,cypher]
+----
+CALL apoc.load.html("test.html",{h6:"h6"}, {failSilently: 'WITH_LIST'});
+----
+
+.Results
+[opts="header"]
+|===
+| Output
+a|
+[source,json]
+----
+{
+  "errorList": [
+    "<h6 i d="error">test</h6>"
+  ],
+  "h6": [
+    {
+      "attributes": {
+        "id": "correct"
+      },
+      "text": "test",
+      "tagName": "h6"
+    }
+  ]
+}
+----
+|===
+
+or with failSilently `WITH_LOG` (note that will be created a `log.warn("Error during parsing element: <h6 i d="error">test</h6>")` ):
+
+[source,cypher]
+----
+CALL apoc.load.html("test.html",{h6:"h6"}, {failSilently: 'WITH_LOG'});
+----
+
+.Results
+[opts="header"]
+|===
+| Output
+a|
+[source,json]
+----
+{
+  "h6": [
+    {
+      "attributes": {
+        "id": "correct"
+      },
+      "text": "test",
+      "tagName": "h6"
+    }
+  ]
+}
+----
+|===

--- a/docs/asciidoc/modules/ROOT/partials/usage/config/apoc.load.html.adoc
+++ b/docs/asciidoc/modules/ROOT/partials/usage/config/apoc.load.html.adoc
@@ -6,5 +6,5 @@ The procedure support the following config parameters:
 | name | type | default | description
 | charset | String | "UTF-8" | the character set of the page being scraped
 | baseUri | String | "" | Base URI used to resolve relative paths
-| failSilently | boolean | false | Return the list of the values, except for values not correctly parsed
+| failSilently | boolean | false | Returns the list of the values, except for values not correctly parsed
 |===

--- a/docs/asciidoc/modules/ROOT/partials/usage/config/apoc.load.html.adoc
+++ b/docs/asciidoc/modules/ROOT/partials/usage/config/apoc.load.html.adoc
@@ -6,5 +6,5 @@ The procedure support the following config parameters:
 | name | type | default | description
 | charset | String | "UTF-8" | the character set of the page being scraped
 | baseUri | String | "" | Base URI used to resolve relative paths
-| failSilently | boolean | false | Return an empty list if a value to return doesn't parse correctly, for example `"a": []`
+| failSilently | boolean | false | Return the list of the values, except for values not correctly parsed
 |===

--- a/docs/asciidoc/modules/ROOT/partials/usage/config/apoc.load.html.adoc
+++ b/docs/asciidoc/modules/ROOT/partials/usage/config/apoc.load.html.adoc
@@ -6,5 +6,5 @@ The procedure support the following config parameters:
 | name | type | default | description
 | charset | String | "UTF-8" | the character set of the page being scraped
 | baseUri | String | "" | Base URI used to resolve relative paths
-| failSilently | boolean | false | Return an empty map if a value to return doesn't parse correctly, for example `"a": {}`
+| failSilently | boolean | false | Return an empty list if a value to return doesn't parse correctly, for example `"a": []`
 |===

--- a/docs/asciidoc/modules/ROOT/partials/usage/config/apoc.load.html.adoc
+++ b/docs/asciidoc/modules/ROOT/partials/usage/config/apoc.load.html.adoc
@@ -6,4 +6,5 @@ The procedure support the following config parameters:
 | name | type | default | description
 | charset | String | "UTF-8" | the character set of the page being scraped
 | baseUri | String | "" | Base URI used to resolve relative paths
+| failSilently | boolean | false | Return an empty map if a value to return doesn't parse correctly, for example `"a": {}`
 |===

--- a/docs/asciidoc/modules/ROOT/partials/usage/config/apoc.load.html.adoc
+++ b/docs/asciidoc/modules/ROOT/partials/usage/config/apoc.load.html.adoc
@@ -6,5 +6,5 @@ The procedure support the following config parameters:
 | name | type | default | description
 | charset | String | "UTF-8" | the character set of the page being scraped
 | baseUri | String | "" | Base URI used to resolve relative paths
-| failSilently | boolean | false | Returns the list of the values, except for values not correctly parsed
+| failSilently | Enum [FALSE, WITH_LOG, WITH_LIST] | FALSE | If the parse fails with one or more elements, using `FALSE` it throws a `RuntimeException`, using `WITH_LOG` a `log.warn` is created for each incorrect item and using `WITH_LIST` an `errorList` key is added to the result with the failed tags.
 |===

--- a/full/src/main/java/apoc/load/LoadHtml.java
+++ b/full/src/main/java/apoc/load/LoadHtml.java
@@ -15,6 +15,8 @@ import org.neo4j.procedure.Description;
 import org.neo4j.procedure.Name;
 import org.neo4j.procedure.Procedure;
 
+import java.io.FileNotFoundException;
+import java.io.IOException;
 import java.util.*;
 import java.util.stream.Stream;
 
@@ -44,14 +46,19 @@ public class LoadHtml {
 
             Map<String, Object> output = new HashMap<>();
 
-            query.keySet().stream().forEach(key -> {
-                Elements elements = document.select(query.get(key));
+            query.keySet().forEach(key -> {
+                try {
+                    Elements elements = document.select(query.get(key));
 
-                output.put(key, getElements(elements, config));
+                    output.put(key, getElements(elements, config));
+
+                } catch (Exception e) {
+                    output.put(key, Collections.emptyMap());
+                }
             });
 
-            return Stream.of( new MapResult(output) );
-        } catch(Exception e){
+            return Stream.of(new MapResult(output));
+        } catch(Exception e) {
             throw new RuntimeException("Can't read the HTML from: "+ url);
         }
     }

--- a/full/src/main/java/apoc/load/LoadHtml.java
+++ b/full/src/main/java/apoc/load/LoadHtml.java
@@ -49,8 +49,7 @@ public class LoadHtml {
 
             query.keySet().forEach(key -> {
                         Elements elements = document.select(query.get(key));
-                        List<Map<String, Object>> elementsParsed = getElements(elements, config);
-                        output.put(key, elementsParsed == null ? Collections.emptyMap() : elementsParsed);
+                        output.put(key, getElements(elements, config));
             });
 
             return Stream.of(new MapResult(output) );
@@ -86,10 +85,11 @@ public class LoadHtml {
 
                     elementList.add(result);
             } catch (Exception e) {
+                final String parseError = "Error during parsing element: " + element;
                 if (failSilently) {
-                    return null;
+                    log.warn(parseError);
                 } else {
-                    throw new RuntimeException("Error during parsing element: " + element);
+                    throw new RuntimeException(parseError);
                 }
             }
         }

--- a/full/src/main/java/apoc/load/LoadHtml.java
+++ b/full/src/main/java/apoc/load/LoadHtml.java
@@ -15,8 +15,6 @@ import org.neo4j.procedure.Description;
 import org.neo4j.procedure.Name;
 import org.neo4j.procedure.Procedure;
 
-import java.io.FileNotFoundException;
-import java.io.IOException;
 import java.util.*;
 import java.util.stream.Stream;
 
@@ -57,7 +55,7 @@ public class LoadHtml {
                 }
             });
 
-            return Stream.of(new MapResult(output));
+            return Stream.of( new MapResult(output) );
         } catch(Exception e) {
             throw new RuntimeException("Can't read the HTML from: "+ url);
         }

--- a/full/src/main/java/apoc/load/LoadHtml.java
+++ b/full/src/main/java/apoc/load/LoadHtml.java
@@ -15,6 +15,9 @@ import org.neo4j.procedure.Description;
 import org.neo4j.procedure.Name;
 import org.neo4j.procedure.Procedure;
 
+import java.io.FileNotFoundException;
+import java.io.UnsupportedEncodingException;
+
 import java.util.*;
 import java.util.stream.Stream;
 
@@ -34,9 +37,9 @@ public class LoadHtml {
         return readHtmlPage(url, query, config);
     }
 
-    private Stream<MapResult> readHtmlPage(String url, Map<String, String> query, Map<String, Object> config){
+    private Stream<MapResult> readHtmlPage(String url, Map<String, String> query, Map<String, Object> config) {
+        String charset = config.getOrDefault("charset", "UTF-8").toString();
         try {
-            String charset = config.getOrDefault("charset", "UTF-8").toString();
             // baseUri is used to resolve relative paths
             String baseUri = config.getOrDefault("baseUri", "").toString();
 
@@ -45,42 +48,50 @@ public class LoadHtml {
             Map<String, Object> output = new HashMap<>();
 
             query.keySet().forEach(key -> {
-                try {
-                    Elements elements = document.select(query.get(key));
-
-                    output.put(key, getElements(elements, config));
-
-                } catch (Exception e) {
-                    output.put(key, Collections.emptyMap());
-                }
+                        Elements elements = document.select(query.get(key));
+                        List<Map<String, Object>> elementsParsed = getElements(elements, config);
+                        output.put(key, elementsParsed == null ? Collections.emptyMap() : elementsParsed);
             });
 
-            return Stream.of( new MapResult(output) );
+            return Stream.of(new MapResult(output) );
+        } catch (FileNotFoundException e) {
+            throw new RuntimeException("File not found from: " + url);
+        } catch(UnsupportedEncodingException e) {
+            throw new RuntimeException("Unsupported charset: " + charset);
         } catch(Exception e) {
-            throw new RuntimeException("Can't read the HTML from: "+ url);
+            throw new RuntimeException("Can't read the HTML from: "+ url, e);
         }
     }
 
     private List<Map<String, Object>> getElements(Elements elements, Map<String, Object> config) {
+        boolean failSilently = Util.toBoolean(config.getOrDefault("failSilently", false));
         List<Map<String, Object>> elementList = new ArrayList<>();
 
         for (Element element : elements) {
-            Map<String, Object> result = new HashMap<>();
-            if(element.attributes().size() > 0) result.put("attributes", getAttributes(element));
-            if(!element.data().isEmpty()) result.put("data", element.data());
-            if(!element.val().isEmpty()) result.put("value", element.val());
-            if(!element.tagName().isEmpty()) result.put("tagName", element.tagName());
+            try {
+                    Map<String, Object> result = new HashMap<>();
+                    if(element.attributes().size() > 0) result.put("attributes", getAttributes(element));
+                    if(!element.data().isEmpty()) result.put("data", element.data());
+                    if(!element.val().isEmpty()) result.put("value", element.val());
+                    if(!element.tagName().isEmpty()) result.put("tagName", element.tagName());
 
-            if ( Util.toBoolean( config.getOrDefault("children", false) ) ) {
-                if(element.hasText())  result.put("text", element.ownText());
+                    if (Util.toBoolean(config.getOrDefault("children", false))) {
+                        if(element.hasText()) result.put("text", element.ownText());
 
-                result.put("children", getElements(element.children(), config));
+                        result.put("children", getElements(element.children(), config));
+                    }
+                    else {
+                        if(element.hasText()) result.put("text", element.text());
+                    }
+
+                    elementList.add(result);
+            } catch (Exception e) {
+                if (failSilently) {
+                    return null;
+                } else {
+                    throw new RuntimeException("Error during parsing element: " + element);
+                }
             }
-            else {
-                if(element.hasText()) result.put("text", element.text());
-            }
-
-            elementList.add(result);
         }
 
         return elementList;

--- a/full/src/main/java/apoc/load/LoadHtml.java
+++ b/full/src/main/java/apoc/load/LoadHtml.java
@@ -24,6 +24,11 @@ import java.util.stream.Stream;
 @Extended
 public class LoadHtml {
 
+    // public for test purpose
+    public static final String KEY_ERROR = "errorList";
+
+    private enum FailSilently { FALSE, WITH_LOG, WITH_LIST }
+
     @Context
     public GraphDatabaseService db;
 
@@ -46,11 +51,15 @@ public class LoadHtml {
             Document document = Jsoup.parse(Util.openInputStream(url, null, null), charset, baseUri);
 
             Map<String, Object> output = new HashMap<>();
+            List<String> errorList = new ArrayList<>();
 
             query.keySet().forEach(key -> {
                         Elements elements = document.select(query.get(key));
-                        output.put(key, getElements(elements, config));
+                        output.put(key, getElements(elements, config, errorList));
             });
+            if (!errorList.isEmpty()) {
+                output.put(KEY_ERROR, errorList);
+            }
 
             return Stream.of(new MapResult(output) );
         } catch (FileNotFoundException e) {
@@ -62,8 +71,9 @@ public class LoadHtml {
         }
     }
 
-    private List<Map<String, Object>> getElements(Elements elements, Map<String, Object> config) {
-        boolean failSilently = Util.toBoolean(config.getOrDefault("failSilently", false));
+    private List<Map<String, Object>> getElements(Elements elements, Map<String, Object> config, List<String> errorList) {
+
+        FailSilently failConfig = FailSilently.valueOf((String) config.getOrDefault("failSilently", "FALSE"));
         List<Map<String, Object>> elementList = new ArrayList<>();
 
         for (Element element : elements) {
@@ -77,7 +87,7 @@ public class LoadHtml {
                     if (Util.toBoolean(config.getOrDefault("children", false))) {
                         if(element.hasText()) result.put("text", element.ownText());
 
-                        result.put("children", getElements(element.children(), config));
+                        result.put("children", getElements(element.children(), config, errorList));
                     }
                     else {
                         if(element.hasText()) result.put("text", element.text());
@@ -86,10 +96,15 @@ public class LoadHtml {
                     elementList.add(result);
             } catch (Exception e) {
                 final String parseError = "Error during parsing element: " + element;
-                if (failSilently) {
-                    log.warn(parseError);
-                } else {
-                    throw new RuntimeException(parseError);
+                switch (failConfig) {
+                    case WITH_LOG:
+                        log.warn(parseError);
+                        break;
+                    case WITH_LIST:
+                        errorList.add(element.toString());
+                        break;
+                    default:
+                        throw new RuntimeException(parseError);
                 }
             }
         }

--- a/full/src/test/java/apoc/load/LoadHtmlTest.java
+++ b/full/src/test/java/apoc/load/LoadHtmlTest.java
@@ -1,6 +1,7 @@
 package apoc.load;
 
 import apoc.util.TestUtil;
+import org.apache.commons.lang.exception.ExceptionUtils;
 import org.junit.Before;
 import org.junit.Rule;
 import org.junit.Test;
@@ -14,9 +15,11 @@ import java.util.List;
 import java.util.Map;
 
 import static apoc.util.MapUtil.map;
-import static apoc.util.TestUtil.testFail;
+import static apoc.util.TestUtil.testCall;
 import static apoc.util.TestUtil.testResult;
 import static java.util.Arrays.asList;
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.containsString;
 import static org.junit.Assert.assertEquals;
 import static org.junit.Assert.assertFalse;
 
@@ -39,6 +42,9 @@ public class LoadHtmlTest {
 
     @Rule
     public DbmsRule db = new ImpermanentDbmsRule();
+
+//    @Rule
+//    public ExpectedException exceptionRule = ExpectedException.none();
 
     @Before
     public void setup() {
@@ -126,10 +132,11 @@ public class LoadHtmlTest {
     }
 
     @Test
-    public void testQueryFailsSilently() {
+    public void testQueryWithFailsSilently() {
         Map<String, Object> query = map("a", "a", "h2", "h2");
 
-        testResult(db, "CALL apoc.load.html($url,$query)", map("url", new File("src/test/resources/wikipedia.html").toURI().toString(), "query", query),
+        testResult(db, "CALL apoc.load.html($url,$query, {failSilently: true})",
+                map("url", new File("src/test/resources/wikipedia.html").toURI().toString(), "query", query),
                 result -> {
                     Map<String, Object> row = result.next();
                     Map<String, Object> value = (Map<String, Object>) row.get("value");
@@ -140,12 +147,47 @@ public class LoadHtmlTest {
 
     }
 
-    @Test
+    @Test(expected = QueryExecutionException.class)
+    public void testQueryWithoutFailsSilently() {
+        final String url = new File("src/test/resources/wikipedia.html").toURI().toString();
+        try {
+            Map<String, Object> query = map("a", "a", "h2", "h2");
+            testCall(db, "CALL apoc.load.html($url,$query)", map("url", url, "query", query), (r) -> {});
+        } catch (Exception e) {
+            Throwable except = ExceptionUtils.getRootCause(e);
+            String expectedMessage = "Error during parsing element: ";
+            assertThat(except.getMessage(), containsString(expectedMessage));
+            throw e;
+        }
+    }
+
+    @Test(expected = QueryExecutionException.class)
     public void testQueryWithExceptionIfIncorrectUrl() {
 
-        testFail(db,
-                "CALL apoc.load.html('" + new File("src/test/resources/wikipedia1.html").toURI().toString() + "',{a:'a'})",
-                QueryExecutionException.class);
+        final String filePath = new File("src/test/resources/wikipedia1.html").toURI().toString();
+        try {
+            String query = "CALL apoc.load.html('" + filePath + "',{a:'a'})";
+            testCall(db, query, (r) -> {});
+        } catch (Exception e) {
+            Throwable except = ExceptionUtils.getRootCause(e);
+            String expectedMessage = "File not found from: " + filePath;
+            assertEquals(expectedMessage, except.getMessage());
+            throw e;
+        }
+    }
 
+    @Test(expected = QueryExecutionException.class)
+    public void testQueryWithExceptionIfIncorrectCharset() {
+        String invalidCharset = "notValid";
+
+        try {
+            String query = "CALL apoc.load.html('" + new File("src/test/resources/wikipedia.html").toURI().toString() + "',{a:'a'}, {charset: '" + invalidCharset + "'})";
+            testCall(db, query, (r) -> {});
+        } catch (Exception e) {
+            Throwable except = ExceptionUtils.getRootCause(e);
+            String expectedMessage = "Unsupported charset: " + invalidCharset;
+            assertEquals(expectedMessage, except.getMessage());
+            throw e;
+        }
     }
 }

--- a/full/src/test/java/apoc/load/LoadHtmlTest.java
+++ b/full/src/test/java/apoc/load/LoadHtmlTest.java
@@ -43,9 +43,6 @@ public class LoadHtmlTest {
     @Rule
     public DbmsRule db = new ImpermanentDbmsRule();
 
-//    @Rule
-//    public ExpectedException exceptionRule = ExpectedException.none();
-
     @Before
     public void setup() {
         TestUtil.registerProcedure(db, LoadHtml.class);
@@ -133,7 +130,7 @@ public class LoadHtmlTest {
 
     @Test
     public void testQueryWithFailsSilently() {
-        Map<String, Object> query = map("a", "a", "h2", "h2");
+        Map<String, Object> query = map("h2", "h2","a", "a");
 
         testResult(db, "CALL apoc.load.html($url,$query, {failSilently: true})",
                 map("url", new File("src/test/resources/wikipedia.html").toURI().toString(), "query", query),

--- a/full/src/test/java/apoc/load/LoadHtmlTest.java
+++ b/full/src/test/java/apoc/load/LoadHtmlTest.java
@@ -18,8 +18,6 @@ import static apoc.util.MapUtil.map;
 import static apoc.util.TestUtil.testCall;
 import static apoc.util.TestUtil.testResult;
 import static java.util.Arrays.asList;
-import static org.hamcrest.MatcherAssert.assertThat;
-import static org.hamcrest.Matchers.containsString;
 import static org.junit.Assert.assertEquals;
 import static org.junit.Assert.assertFalse;
 
@@ -130,18 +128,19 @@ public class LoadHtmlTest {
 
     @Test
     public void testQueryWithFailsSilently() {
-        Map<String, Object> query = map("h2", "h2","a", "a");
+        Map<String, Object> query = map("h2", "h2","a", "a", "invalid", "invalid");
 
         testResult(db, "CALL apoc.load.html($url,$query, {failSilently: true})",
                 map("url", new File("src/test/resources/wikipedia.html").toURI().toString(), "query", query),
                 result -> {
                     Map<String, Object> row = result.next();
                     Map<String, Object> value = (Map<String, Object>) row.get("value");
-                    assertEquals(Collections.emptyMap(), value.get("a"));
+                    // number of <a> tags in html file minus the incorrect tag
+                    assertEquals(107, ((List) value.get("a")).size());
+                    assertEquals(Collections.emptyList(), value.get("invalid"));
                     assertEquals(List.of(RESULT_QUERY_H2).toString().trim(), value.get("h2").toString().trim());
                     assertFalse(result.hasNext());
                 });
-
     }
 
     @Test(expected = QueryExecutionException.class)
@@ -152,8 +151,8 @@ public class LoadHtmlTest {
             testCall(db, "CALL apoc.load.html($url,$query)", map("url", url, "query", query), (r) -> {});
         } catch (Exception e) {
             Throwable except = ExceptionUtils.getRootCause(e);
-            String expectedMessage = "Error during parsing element: ";
-            assertThat(except.getMessage(), containsString(expectedMessage));
+            String expectedMessage = "Error during parsing element: <a hre f=\"#cite_ref-Simpson2018_1-0\"><sup><i><b>a</b></i></sup></a>";
+            assertEquals(expectedMessage, except.getMessage());
             throw e;
         }
     }

--- a/full/src/test/java/apoc/load/LoadHtmlTest.java
+++ b/full/src/test/java/apoc/load/LoadHtmlTest.java
@@ -14,12 +14,14 @@ import java.util.Collections;
 import java.util.List;
 import java.util.Map;
 
+import static apoc.load.LoadHtml.KEY_ERROR;
 import static apoc.util.MapUtil.map;
 import static apoc.util.TestUtil.testCall;
 import static apoc.util.TestUtil.testResult;
 import static java.util.Arrays.asList;
 import static org.junit.Assert.assertEquals;
 import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertNull;
 
 public class LoadHtmlTest {
 
@@ -37,6 +39,10 @@ public class LoadHtmlTest {
             "{text=Legacy[edit], tagName=h2}, " +
             "{text=References[edit], tagName=h2}, " +
             "{text=Navigation menu, tagName=h2}");
+
+    private static final String INVALID_PATH = new File("src/test/resources/wikipedia1.html").toURI().toString();
+    private static final String VALID_PATH = new File("src/test/resources/wikipedia.html").toURI().toString();
+    private static final String INVALID_CHARSET = "notValid";
 
     @Rule
     public DbmsRule db = new ImpermanentDbmsRule();
@@ -127,10 +133,10 @@ public class LoadHtmlTest {
     }
 
     @Test
-    public void testQueryWithFailsSilently() {
-        Map<String, Object> query = map("h2", "h2","a", "a", "invalid", "invalid");
+    public void testQueryWithFailsSilentlyWithLog() {
+        Map<String, Object> query = map("a", "a", "invalid", "invalid", "h6", "h6");
 
-        testResult(db, "CALL apoc.load.html($url,$query, {failSilently: true})",
+        testResult(db, "CALL apoc.load.html($url,$query, {failSilently: 'WITH_LOG'})",
                 map("url", new File("src/test/resources/wikipedia.html").toURI().toString(), "query", query),
                 result -> {
                     Map<String, Object> row = result.next();
@@ -138,7 +144,47 @@ public class LoadHtmlTest {
                     // number of <a> tags in html file minus the incorrect tag
                     assertEquals(107, ((List) value.get("a")).size());
                     assertEquals(Collections.emptyList(), value.get("invalid"));
-                    assertEquals(List.of(RESULT_QUERY_H2).toString().trim(), value.get("h2").toString().trim());
+                    assertNull(value.get(KEY_ERROR));
+                    assertFalse(result.hasNext());
+                });
+    }
+
+    @Test
+    public void testQueryWithFailsSilentlyWithList() {
+        Map<String, Object> query = map("a", "a", "invalid", "invalid", "h6", "h6");
+
+        String expectedH6 = "[{attributes={id=correct}, text=test, tagName=h6}, {attributes={id=childIncorrect}, text=incorrecttest, tagName=h6}]";
+
+        testResult(db, "CALL apoc.load.html($url,$query, {failSilently: 'WITH_LIST'})",
+                map("url", new File("src/test/resources/wikipedia.html").toURI().toString(), "query", query),
+                result -> {
+                    Map<String, Object> row = result.next();
+                    Map<String, Object> value = (Map<String, Object>) row.get("value");
+                    // number of <a> tags in html file minus the incorrect tag
+                    assertEquals(107, ((List) value.get("a")).size());
+                    assertEquals(Collections.emptyList(), value.get("invalid"));
+                    assertEquals(expectedH6, value.get("h6").toString().trim());
+                    assertEquals(2, ((List) value.get(KEY_ERROR)).size());
+                    assertFalse(result.hasNext());
+                });
+    }
+
+    @Test
+    public void testQueryWithFailsSilentlyWithListAndChildren() {
+        Map<String, Object> query = map("a", "a", "invalid", "invalid", "h6", "h6");
+
+        String expectedH6 = "[{children=[], attributes={id=correct}, text=test, tagName=h6}, {children=[], attributes={id=childIncorrect}, text=incorrect, tagName=h6}]";
+
+        testResult(db, "CALL apoc.load.html($url,$query, {failSilently: 'WITH_LIST', children: true})",
+                map("url", new File("src/test/resources/wikipedia.html").toURI().toString(), "query", query),
+                result -> {
+                    Map<String, Object> row = result.next();
+                    Map<String, Object> value = (Map<String, Object>) row.get("value");
+                    // number of <a> tags in html file minus the incorrect tag
+                    assertEquals(107, ((List) value.get("a")).size());
+                    assertEquals(Collections.emptyList(), value.get("invalid"));
+                    assertEquals(expectedH6, value.get("h6").toString().trim());
+                    assertEquals(3, ((List) value.get(KEY_ERROR)).size());
                     assertFalse(result.hasNext());
                 });
     }
@@ -159,29 +205,51 @@ public class LoadHtmlTest {
 
     @Test(expected = QueryExecutionException.class)
     public void testQueryWithExceptionIfIncorrectUrl() {
+        testIncorrectUrl("CALL apoc.load.html('" + INVALID_PATH + "',{a:'a'})");
+    }
 
-        final String filePath = new File("src/test/resources/wikipedia1.html").toURI().toString();
+    @Test(expected = QueryExecutionException.class)
+    public void testQueryWithFailsSilentlyWithLogWithExceptionIfIncorrectUrl() {
+        testIncorrectUrl("CALL apoc.load.html('" + INVALID_PATH + "',{failSilently: 'WITH_LOG', a:'a'})");
+    }
+
+    @Test(expected = QueryExecutionException.class)
+    public void testQueryWithFailsSilentlyWithListWithExceptionIfIncorrectUrl() {
+        testIncorrectUrl("CALL apoc.load.html('" + INVALID_PATH + "',{failSilently: 'WITH_LIST', a:'a'})");
+    }
+
+    @Test(expected = QueryExecutionException.class)
+    public void testQueryWithExceptionIfIncorrectCharset() {
+        testIncorrectCharset("CALL apoc.load.html('" + VALID_PATH + "',{a:'a'}, {charset: '" + INVALID_CHARSET + "'})");
+    }
+
+    @Test(expected = QueryExecutionException.class)
+    public void testQueryWithFailsSilentlyWithLogWithExceptionIfIncorrectCharset() {
+        testIncorrectCharset("CALL apoc.load.html('" + VALID_PATH + "',{a:'a'}, {failSilently: 'WITH_LOG', charset: '" + INVALID_CHARSET + "'})");
+    }
+
+    @Test(expected = QueryExecutionException.class)
+    public void testQueryWithFailsSilentlyWithListWithExceptionIfIncorrectCharset() {
+        testIncorrectCharset("CALL apoc.load.html('" + VALID_PATH + "',{a:'a'}, {failSilently: 'WITH_LIST', charset: '" + INVALID_CHARSET + "'})");
+    }
+
+    private void testIncorrectCharset(String query) {
         try {
-            String query = "CALL apoc.load.html('" + filePath + "',{a:'a'})";
             testCall(db, query, (r) -> {});
         } catch (Exception e) {
             Throwable except = ExceptionUtils.getRootCause(e);
-            String expectedMessage = "File not found from: " + filePath;
+            String expectedMessage = "Unsupported charset: " + INVALID_CHARSET;
             assertEquals(expectedMessage, except.getMessage());
             throw e;
         }
     }
 
-    @Test(expected = QueryExecutionException.class)
-    public void testQueryWithExceptionIfIncorrectCharset() {
-        String invalidCharset = "notValid";
-
+    private void testIncorrectUrl(String query) {
         try {
-            String query = "CALL apoc.load.html('" + new File("src/test/resources/wikipedia.html").toURI().toString() + "',{a:'a'}, {charset: '" + invalidCharset + "'})";
             testCall(db, query, (r) -> {});
         } catch (Exception e) {
             Throwable except = ExceptionUtils.getRootCause(e);
-            String expectedMessage = "Unsupported charset: " + invalidCharset;
+            String expectedMessage = "File not found from: " + INVALID_PATH;
             assertEquals(expectedMessage, except.getMessage());
             throw e;
         }

--- a/full/src/test/java/apoc/load/LoadHtmlTest.java
+++ b/full/src/test/java/apoc/load/LoadHtmlTest.java
@@ -4,6 +4,7 @@ import apoc.util.TestUtil;
 import org.junit.Before;
 import org.junit.Rule;
 import org.junit.Test;
+import org.neo4j.graphdb.QueryExecutionException;
 import org.neo4j.test.rule.DbmsRule;
 import org.neo4j.test.rule.ImpermanentDbmsRule;
 
@@ -13,6 +14,7 @@ import java.util.List;
 import java.util.Map;
 
 import static apoc.util.MapUtil.map;
+import static apoc.util.TestUtil.testFail;
 import static apoc.util.TestUtil.testResult;
 import static java.util.Arrays.asList;
 import static org.junit.Assert.assertEquals;
@@ -121,5 +123,29 @@ public class LoadHtmlTest {
                     assertEquals("li", firstChild.get("tagName"));
                     assertEquals(1, ((List) firstChild.get("children")).size());
                 });
+    }
+
+    @Test
+    public void testQueryFailsSilently() {
+        Map<String, Object> query = map("a", "a", "h2", "h2");
+
+        testResult(db, "CALL apoc.load.html($url,$query)", map("url", new File("src/test/resources/wikipedia.html").toURI().toString(), "query", query),
+                result -> {
+                    Map<String, Object> row = result.next();
+                    Map<String, Object> value = (Map<String, Object>) row.get("value");
+                    assertEquals(Collections.emptyMap(), value.get("a"));
+                    assertEquals(List.of(RESULT_QUERY_H2).toString().trim(), value.get("h2").toString().trim());
+                    assertFalse(result.hasNext());
+                });
+
+    }
+
+    @Test
+    public void testQueryWithExceptionIfIncorrectUrl() {
+
+        testFail(db,
+                "CALL apoc.load.html('" + new File("src/test/resources/wikipedia1.html").toURI().toString() + "',{a:'a'})",
+                QueryExecutionException.class);
+
     }
 }

--- a/full/src/test/resources/wikipedia.html
+++ b/full/src/test/resources/wikipedia.html
@@ -205,6 +205,9 @@ Transclusion expansion time report (%,ms,calls,template)
 		</div>
 			<div class="portal" role="navigation" id="p-lang" aria-labelledby="p-lang-label">
 			<h3 id="p-lang-label">Languages</h3>
+			<h6 i d="error">test</h6>
+			<h6 id="correct">test</h6>
+			<h6 id="childIncorrect">incorrect<span i d="incorrect">test</span></h6>
 			<div class="body">
 								<ul>
 									</ul>

